Multitenancy support

machisuji opened the pull request:
More/better delayed jobs and other stuff for multitenancy.

WP [#18226](https://community.openproject.org/work_packages/18226)

See also: [openproject-multitenancy PR #14](https://github.com/finnlabs/openproject-multitenancy/pull/14)
See also: local avatars PR [#11](https://github.com/finnlabs/openproject-local_avatars/pull/11)

ulferts committed:
always set current_user to nil

This will avoid AnymousUsers from being set in one spec and then that
same user being removed in the db but staying set in the class instance
variable.

machisuji committed:
use IDs in CopyProjectJob; send mail synchronously

also moved the job into the workers directory

machisuji committed:
created delayed job object for wp update emails

machisuji committed:
allow + in user logins

machisuji committed:
make force_ssl configurable (default false)

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://feature-dev-multitenancy.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://feature-dev-multitenancy.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

myabc commented on app/models/user.rb:
```diff
@@ -144,7 +144,7 @@ def self.create_blocked_scope(blocked)
   validates_uniqueness_of :mail, allow_blank: true, case_sensitive: false
   # Login must contain letters, numbers, underscores only
-  validates_format_of :login, with: /\A[a-z0-9_\-@\.]*\z/i
+  validates_format_of :login, with: /\A[a-z0-9_\-@\.+]*\z/i
```
@houndci what do you know about sexiness?

0xF013 replied:
@myabc doggy-style

myabc replied:
@0xF013 no comment :poodle:

0xF013 replied:
:dog:

linki replied:
:heart_eyes_cat:

machisuji committed:
fixed broken wp spec

the current user within the shared example is a
different one to the one (anonymous from the outmost spec scope) than the one defined in the 'user with permission to view time entries' context, which, I believe, is unintended and also happens to break the spec with the changes to the journal observer.

machisuji committed:
fixed toppler [1] by not using anon user to update

[1] a test that may or may not fail depending on the time of day, the weather, or the kind of pants you wear.

machisuji committed:
fixed spec (+ is now allowed)

machisuji committed:
make sure data is saved after creation already too

machisuji committed:
fixed mail handler test

machisuji committed:
reset current user between tests

linki commented:
lgtm :+1:

machisuji commented:
@myabc I [put it](https://github.com/opf/openproject/commit/a08cac4741e472d555e56163d90f08e476bb52b0) into `spec/support` because I didn't want to stuff any more code into the `spec_helper`. Dunno if this is the right thing to do. If there are objections I can also just add the 3 lines removed from the `copy_project_job_spec` to the `spec_helper`.

ulferts commented on spec/support/global_state_reset.rb:
```diff
@@ -0,0 +1,22 @@
+
+RSpec.configure do |config|
+  config.after(:suite) do
+    OpenProject::GlobalStateReset.reset_global_state!
```
@machisuji This duplicates what we have in https://github.com/opf/openproject/blob/dev/spec/spec_helper.rb#L197. Additionally, it sets User.current to an AnonymousUser which will be left in the DB. I propose setting this to nil as well and remove the duplicity with the spec_helper.

machisuji replied:
@ulferts Ah right. ^^ Didn't see that. In that case I'll just drop the commit altogether.

myabc replied:
@ulferts you simplified the global state cleanup on the Rails 4 branch: https://github.com/opf/openproject/commit/581a547e5554fba150a57c8d2bade071439b2c21 would that patch help here?

ulferts replied:
@myabc good pointer!

machisuji replied:
We don't technically need anything after https://github.com/opf/openproject/commit/f7662130a96c1f36333a41db68b2c0d98e49750d. Originally I just added the `config.after(:suite) { User.current = ...` to `copy_project_job_spec.rb` to get rid of the "lazyness warning".

But I do think @ulferts's refactoring is good and would allow me and everyone else not to bother with that anymore. So I would like to go ahead and just cherry-pick that commit then.

myabc replied:
@machisuji :+1: for :cherries:-picking

linki commented:
thx @machisuji
